src/nupic/utils/SlidingWindow.hpp

```diff
@@ -0,0 +1,191 @@
+        :return bool if some value has been dropped (and updated as 
+          droppedValue) 
+      */
+      bool append(T newValue, T& droppedValue) {
+        //only in this case we drop oldest; this happens always after
+        //first maxCap steps ; must be checked before append()
+        bool isFull = (buffer_.size()==maxCapacity);
+        if(isFull) {
+          droppedValue = buffer_[idxNext_];
+        }
+        append(newValue);
+        return isFull;
+      }
+
+
+      /**
+        :return unordered content (data ) of this sl. window; 
+          call getLinearizedData() if you need them oredered from 
+          oldest->newest
+        This direct access method is fast.
+      */
+      const std::vector<T>& getData() const {
+        return buffer_;
+      }
+
+
+      /** linearize method for the internal buffer; this is slower than 
+        the pure getData() but ensures that the data are ordered (oldest at
+        the beginning, newest at the end of the vector
+        This handles case of |5,6;1,2,3,4| => |1,2,3,4,5,6|
+        :return new linearized vector
+      */
+      std::vector<T> getLinearizedData() const {
+        std::vector<T> lin;
+        lin.reserve(buffer_.size());
+
+        //insert the "older" part at the beginning
+        lin.insert(std::begin(lin), std::begin(buffer_) + idxNext_, std::end(buffer_));
+        //append the "newer" part to the end of the constructed vect
+        lin.insert(std::end(lin), std::begin(buffer_), std::begin(buffer_) + idxNext_);
+        return lin;
+      }
+
+
+      bool operator==(const SlidingWindow& r2) const {
+        return ((this->size() == r2.size()) && (this->maxCapacity == r2.maxCapacity) &&
+          (this->getData()== r2.getData()) && getLinearizedData() == r2.getLinearizedData());
+        //FIXME review the ==, on my machine it randomly passes/fails the test!
+      }
+
+
+      bool operator!=(const SlidingWindow& r2) const {
+        return !operator==(r2);
+      }
+
+
+      /** operator[] provides fast access to the elements indexed relatively
+        to the oldest element. So slidingWindow[0] returns oldest element, 
+        slidingWindow[size()] returns the newest. 
+      :param UInt index - index/offset from the oldest element, values 0..size()
+      :return T - i-th oldest value in the buffer
+      :throws 0<=index<=size()
+      */ 
+      T& operator[](UInt index) {
+        NTA_ASSERT(index <= size());
+        NTA_ASSERT(size() > 0);
+        //get last updated position, "current"+index(offset)
+        //avoid calling getLinearizeData() as it involves copy()
+        if (size() == maxCapacity) {
+          return &buffer_[(idxNext_ + index) % maxCapacity];
+        } else {
+          return &buffer_[index];
+        }
+      }
+
+
+      const T& operator[](UInt index) const {
+        return this->operator[](index); //call the overloaded operator[] above
+      }
+
+
+      std::ostream& operator<<(std::ostream& os) {
```

Comment:
The problem with having this method is that it makes this class less reusable -- now T must support the `<<` operator, otherwise the code won't compile. This is probably why vectors don't support `<<`. It's better to have callers do this manually, and to focus on making it easy for callers to do this (e.g. by having `begin()` and `end()` iterators).